src/Storage/Notmuch.hs

```diff
@@ -1,39 +1,85 @@
 {-# LANGUAGE FlexibleContexts #-}
 {-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE LambdaCase #-}
 -- | module for integrating notmuch within purebred
 module Storage.Notmuch where
 
 import Types (NotmuchMail(..))
 import Notmuch
 import Notmuch.Search
 
-import           Data.Maybe     (fromMaybe)
-import qualified Data.Vector    as Vec
-import           System.Process (readProcess)
+import Control.Monad (void)
+import Control.Monad.IO.Class (liftIO)
+import Control.Monad.Except (runExceptT)
+import Data.Traversable (traverse)
+import Data.List (union)
+import Data.Maybe (fromMaybe)
+import Data.Bifunctor (first)
+import qualified Data.Vector as Vec
+import System.Process (readProcess)
+import Control.Exception (bracket)
 import qualified Data.Text as T
-import Data.Text.Encoding (decodeUtf8)
-import Types (NotmuchSettings, nmDatabase, mailTags)
+import Data.Text.Encoding (decodeUtf8, encodeUtf8)
+import Types (NotmuchSettings, nmDatabase, mailId, mailTags)
 import Control.Lens.Getter (view)
+import Control.Lens.Setter (over)
 
 
 -- | creates a vector of parsed mails from a not much search
 -- Note, that at this point in time only free form searches are supported. Also,
 -- we filter out the tag which we use to mark mails as new mails
-getMessages :: T.Text -> NotmuchSettings FilePath -> IO (Vec.Vector NotmuchMail)
-getMessages s settings = do
-  db' <- databaseOpen (view nmDatabase settings)
-  case db' of
-    Left status -> do
-        error $ show status
-    Right db -> do
-        q <- query db (FreeForm $ T.unpack s)
-        msgs <- messages q
-        mails <- mapM messageToMail msgs
-        return $ Vec.fromList mails
+getMessages :: T.Text
+            -> NotmuchSettings FilePath
+            -> IO (Either String (Vec.Vector NotmuchMail))
+getMessages s settings =
+  first show <$>
+  bracket (runExceptT (databaseOpenReadOnly (view nmDatabase settings))
+           >>= either (error . show) pure)
```

Comment:
This is still not right IMHO... it would currently crash the application if I wouldn't be able the open the database properly. Arguably that's pretty bad in itself, but perhaps the application shouldn't crash. Same for the function which opens the database RW for tagging. Any suggestions or leave it for now and keep a backlog item.

Reply:
Leave for now.